steps/11_transformer_refactor/step.md

```diff
@@ -0,0 +1,158 @@
+# refactor of transformers interface
+
+Contributors: @fkiraly, 
+
+## High-level summary 
+
+### The Aim
+
+The transformers interface has grown organically and is difficult to use, interface, and extend.
+
+This proposal tries to alleviate the problem by removing mixins from transformers and using the new tag system to provide a unified and easy-to-use interface.
+
+### The proposed solution
+
+Our proposed solution follows principles used in the refactor of the forecasters:
+
+* methods are separated into outer `fit`/`transform` and inner `_fit`/`_transform` etc
+* the four mixins are removed, and replaced by tags
+* the four transformer mixins are united under a single interface
+* addition of `update` methods for transformers, to be usable with forecaster `update`
+
+The unification introduces type inhomogeneity in the `transform` method - this is addressed by allowing both `Series` and `Panel` inputs and outputs, and dropping the `Z` argument in favour of `X` (keeping the `Z` as a downwards compatible alias).
+
+## Refactor design: transformers
+
+We proceed outlining the refactor interface.
+
+### Conceptual model: types of transformers
+
+Conceptually, we change our view of transformer types. Before the refactor, these weree:
+
+* series-to-primitives
+* panel-to-tabular
+* series-to-series
+* panel-to-panel
+* series-to-panel
+
+Where series-to-panel occurs only in two places currently and not formalised as a transformer type:
+* sliding window forecasting (in `update_predict`)
+* epoching, in `sktime-neuro`
```

Comment:
Hi @fkiraly, I'm not entirely sure whether we should handle the epoching as a transformer. (not a strong opinion just bouncing off my thoughts:) What the use case here usually requires is that we can pass a time series and one or more labels, e.g. "right" and "left" and an interval e.g. [-0.5 1]. And then this function/transformer cuts out all the parts of the data where these events occurred. It then returns panel data and a labels vector (describing each instance of the panel). So the return structure would be similar to the one you get when loading a toy dataset in sklearn (it simply returns X and y). And it could then further be used just like a normal classification problem. The critical point I see is that a transformer would not be able to return the labels vector y but would require it to be fixed in the beginning. Which I think is a bit unhandy.

I see two options of forcing this use case into a transformer: 
(1) We could make the user specify only one label. Then the label of each instance in the panel would be clear. And the user would then have to apply this transformer per label, e.g. first cutting out all the "right" trials and then cutting out all the "left" trials. Here we would however have the risk that the user (accidentally) applies different panel-to-panel transformers to the two generated panels afterwards, thereby classifying what happened to the data afterwards rather than purely classifying the label.  
 
(2) We could require the labels vector (y) (indicating which instance of the panel is supposed to be of which label afterwards) to be fixed beforehand. This would basically require a function `create_labels_vector` that creates the labels vector based on the labels we want to cut out and an annotation object (currently a `pd.DataFrame` that contains events and onsets). So instead of calling simply one function we would require the user to first create the labels vector and then to `fit_transform()` the series.

Reply:
This is how I see it would be part of the framework:

* it would be a "series-to-panel" transformer which takes a `y` that is of segmentation scitype.. 
* the segments (in your diction: intervals, e.g., [-0.5, 1]) would be contained in `y`, e.g., as an `IntervalIndex` or in some other encoding. 
* the series would be passed in `X`. The return of transform is a panel which contains all the "cut out" sub-sequences.

Reply:
so sth like this

```python
mytrafo = SegmentCutter()
y = pd.IntervalIndex.from_arrays([-0.5, 10, 14], [1, 12, 15])
epochs = mytrafo.fit_transform(X, y)
```

now `epochs` is a panel which contains cut parts form `X`, in intervals [-0.5, 1], [10,12], and [14,15].

Reply:
But how would you know the labels of the instances in epochs? You'd have to have an additional function that creates the labels vector from y in order to train a classifier, right?

Reply:
ah, I see - you have labels!
Well, in that case, you would start with a labelled segmentation, let's say `Y`, which is a `pd.Series` with a `pd.IntervalIndex`, and entries are labels.
You could now pass `y=Y.index` to the `SegmentCutter` as above, and `Y.reset_index()` would give you the labels - unless you allow panels which can also have a `pd.intervalIndex` as instance index, in which case `epochs` and `Y` already would have the same index and could be used in a classifier.